File: Merge Sort - GFG/merge-sort.java
//{ Driver Code Starts
import java.util.*;

class Merge_Sort
{
    //method to print the elements of the array
	static void printArray(int arr[])
    {
        StringBuffer sb=new StringBuffer("");
        int n = arr.length;
        for (int i=0; i<n; ++i)
            sb.append(arr[i]+" ");
        System.out.println(sb.toString());
    }

    

	public static void main(String args[])
	{
	    //taking input using Scanner class
		Scanner sc = new Scanner(System.in);
		
		//taking testcases
		int T = sc.nextInt();
		while(T>0)
		{
		    //taking elements count
			int n = sc.nextInt();
			
			//creating an object of class Merge_Sort
			Merge_Sort ms = new Merge_Sort();
			
			//creating an array of size n
			int arr[] = new int[n];
			
			//adding elements to the array
			for(int i=0;i<n;i++)
				arr[i] = sc.nextInt();

            
			Solution g = new Solution();
			
			//calling the method mergeSort
			g.mergeSort(arr,0,arr.length-1);

            //calling the method printArray
			ms.printArray(arr);
		T--;
		}
	}
}



// } Driver Code Ends


class Solution
{
    void merge(int arr[], int l, int m, int r)
    {
         int len1=m-l+1;
         int len2= r-(m+1)+1;//r-m
         
        int[] arr1 = new int[len1];
        int[] arr2 = new int[len2];
         
         int k=l; // k=orginalMainIndex
         
         for(int i=0;i<len1;i++){
             arr1[i]=arr[k];
             k++;
            
         }
         
         for(int i=0;i<len2;i++){
             arr2[i]=arr[k];
             k++;
        }
        
        k=l;
        
        //merge two sorted array
        
        int index1=0, index2=0;
        
        while(index1<len1 && index2<len2){
            if(arr1[index1]<arr2[index2]){
                arr[k]=arr1[index1];
                k++;
                index1++;
            }
            else{
                arr[k]=arr2[index2];
                k++;
                index2++;
            }
        }
        
        while(index1<len1){
            arr[k]=arr1[index1];
                k++;
                index1++;
        }
        
        while(index2<len2){
             arr[k]=arr2[index2];
                k++;
                index2++;
        }
        
        
         
         
    }
    void mergeSort(int arr[], int l, int r)
    {
        
            if(l>=r){  // base case
                return ;
            }
       
        
        int mid=l+(r-l)/2;
        
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        
        merge(arr,l,mid,r);
        
        
    }
}

File: Merge Sort - GFG/merge-sort.cpp
//{ Driver Code Starts
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;



/* Function to print an array */
void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}


// } Driver Code Ends
class Solution
{
    public:
    void merge(int arr[], int l, int m, int r)
    {
         // Your code here
         //int n=abs(r-l)+3;
        /// vector<int>b(10000,0);
        int len1=m-l+1;
        int len2=r-m;
        int b[len1+len2+1000000];
         int i=l;
         int k=l;
         int j=m+1;
         while(i<=m && j<=r ){
             if(arr[i]<arr[j]){
                 b[k]=arr[i];
                 i++;
             }
             else{
                 b[k]=arr[j];
                 j++;
             }
             k++;
         }
         
         if(i<=m){
             while(i<=m){
                 b[k]=arr[i];
                 i++;
                 k++;
             }
         }
         else if(j<=r){
             while(j<=r){
                 b[k]=arr[j];
                 j++;
                 k++;
             }
         }
         
         for(int p=l;p<=r;p++){
             arr[p]=b[p];
         }
         
    }
    public:
    void mergeSort(int arr[], int l, int r)
    {
        //code here
        if(l<r){
            
        int mid=(l+r)/2;
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        merge(arr,l,mid,r);
        }
    
    }
};

//{ Driver Code Starts.


int main()
{
    int n,T,i;

    scanf("%d",&T);

    while(T--){
    
    scanf("%d",&n);
    int arr[n+1];
    for(i=0;i<n;i++)
      scanf("%d",&arr[i]);

    Solution ob;
    ob.mergeSort(arr, 0, n-1);
    printArray(arr, n);
    }
    return 0;
}
// } Driver Code Ends